reporter/reporter.go

```diff
@@ -0,0 +1,237 @@
+type WavefrontMetricsReporter interface {
+	Start()
+	Stop()
+}
+
+type reporter struct {
+	sender       wf.Sender
+	application  application.Tags
+	source       string
+	prefix       string
+	addSuffix    bool
+	interval     time.Duration
+	ticker       *time.Ticker
+	registry     *WavefrontRegistry
+	percentiles  []float64              // Percentiles to export from timers and histograms
+	durationUnit time.Duration          // Time conversion unit for durations
+	metrics      map[string]interface{} // for Wavefron specific metrics tyoes, like Histograms
+}
+
+// Option allow WavefrontReporter customization
+type Option func(*reporter)
+
+// Source tag for metrics
+func Source(source string) Option {
+	return func(args *reporter) {
+		args.source = source
+	}
+}
+
+// ReportingInterval chanfe the default 1 minute reporting interval.
+func ReportingInterval(interval time.Duration) Option {
+	return func(args *reporter) {
+		args.interval = interval
+	}
+}
+
+// Registry set the metric registry to report
+func Registry(registry *WavefrontRegistry) Option {
+	return func(args *reporter) {
+		args.registry = registry
+	}
+}
+
+// Prefix for the metrics name
+func Prefix(prefix string) Option {
+	return func(args *reporter) {
+		args.prefix = strings.TrimSuffix(prefix, ".")
+	}
+}
+
+// AddSuffix add a metric suffix based on the metric type ('.count', '.value')
+func AddSuffix(addSuffix bool) Option {
+	return func(args *reporter) {
+		args.addSuffix = addSuffix
+	}
+}
+
+// New create a WavefrontMetricsReporter
+func New(sender wf.Sender, application application.Tags, setters ...Option) WavefrontMetricsReporter {
+	r := &reporter{
+		sender:       sender,
+		application:  application,
+		source:       "mySource",
+		interval:     time.Second * 5,
+		registry:     DefaultWavefrontRegistry,
+		percentiles:  []float64{0.5, 0.75, 0.95, 0.99, 0.999},
+		durationUnit: time.Nanosecond,
+		metrics:      make(map[string]interface{}),
+		addSuffix:    true,
+	}
+
+	for _, setter := range setters {
+		setter(r)
+	}
+
+	return r
+}
+
+// RegisterMetric tag support for metrics.Register()
+func RegisterMetric(key string, metric interface{}, tags map[string]string) {
+	key = EncodeKey(key, tags)
+	DefaultWavefrontRegistry.Register(key, metric)
+}
+
+// GetMetric tag support for metrics.Get()
+func GetMetric(key string, tags map[string]string) interface{} {
+	key = EncodeKey(key, tags)
+	return DefaultWavefrontRegistry.Get(key)
+}
+
+// GetOrRegisterMetric tag support for metrics.GetOrRegister()
+func GetOrRegisterMetric(name string, i interface{}, tags map[string]string) interface{} {
+	key := EncodeKey(name, tags)
+	return DefaultWavefrontRegistry.GetOrRegister(key, i)
+}
+
+// UnregisterMetric tag support for metrics.UnregisterMetric()
+func UnregisterMetric(name string, tags map[string]string) {
+	key := EncodeKey(name, tags)
+	DefaultWavefrontRegistry.Unregister(key)
+}
+
+func (r *reporter) report() {
+	r.registry.Each(func(key string, metric interface{}) {
```

Comment:
The SendMetric calls can return an error. We need a way to surface errors to users (maybe log) so they know when the reporting fails.